allennlp/models/graph_parser.py

```diff
@@ -0,0 +1,341 @@
+                        mask: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Computes the arc and tag loss for an adjacency matrix.
+
+        Parameters
+        ----------
+        attended_arcs : ``torch.Tensor``, required.
+            A tensor of shape (batch_size, sequence_length, sequence_length) used to generate a
+            binary classification decision for whether an edge is present between two words.
+        arc_tag_logits : ``torch.Tensor``, required.
+            A tensor of shape (batch_size, sequence_length, sequence_length, num_tags) used to generate
+            a distribution over edge tags for a given edge.
+        arc_tags : ``torch.Tensor``, required.
+            A tensor of shape (batch_size, sequence_length, sequence_length).
+            The labels for every arc.
+        mask : ``torch.Tensor``, required.
+            A mask of shape (batch_size, sequence_length), denoting unpadded
+            elements in the sequence.
+
+        Returns
+        -------
+        arc_nll : ``torch.Tensor``, required.
+            The negative log likelihood from the arc loss.
+        tag_nll : ``torch.Tensor``, required.
+            The negative log likelihood from the arc tag loss.
+        """
+        float_mask = mask.float()
+        arc_indices = (arc_tags != -1).float()
+        # Make the arc tags not have negative values anywhere
+        # (by default, no edge is indicated with -1).
+        arc_tags = arc_tags * arc_indices
+        arc_nll = self._arc_loss(attended_arcs, arc_indices) * float_mask.unsqueeze(1) * float_mask.unsqueeze(2)
+        # We want the mask for the tags to only include the unmasked words
+        # and we only care about the loss with respect to the gold arcs.
+        tag_mask = float_mask.unsqueeze(1) * float_mask.unsqueeze(2) * arc_indices
+
+        batch_size, sequence_length, _, num_tags = arc_tag_logits.size()
+        original_shape = [batch_size, sequence_length, sequence_length]
+        reshaped_logits = arc_tag_logits.view(-1, num_tags)
+        reshaped_tags = arc_tags.view(-1)
+        tag_nll = self._tag_loss(reshaped_logits, reshaped_tags.long()).view(original_shape) * tag_mask
+
+        valid_positions = tag_mask.sum()
+
+        arc_nll = arc_nll.sum() / valid_positions.float()
+        tag_nll = tag_nll.sum() / valid_positions.float()
+        return arc_nll, tag_nll
+
+    @staticmethod
+    def _greedy_decode(attended_arcs: torch.Tensor,
+                       arc_tag_logits: torch.Tensor,
+                       mask: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Decodes the head and head tag predictions by decoding the unlabeled arcs
+        independently for each word and then again, predicting the head tags of
+        these greedily chosen arcs indpendently. Note that this method of decoding
+        is not guaranteed to produce trees (i.e. there maybe be multiple roots,
```

Comment:
This second sentence probably isn't necessary, as the point is that this is a graph parser.